### PR TITLE
feature(pu): add gfootball academy env and config

### PR DESCRIPTION
**Pull request by puyuan1996**
## Description

- add gfootball academy env: academy_3_vs_1_with_keeper, academy_counterattack_hard, and corresponding masac mappo config.
- add enable_save_replay method for GfootballAcademyEnv and env unittest.
下面是academy_3_vs_1_with_keeper环境，所有智能体采用random action得到的一局的录像。

academy_3_vs_1_with_keeper_episode_0

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/model/template/maqac.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ def __init__(
             agent_obs_shape: Union[int, SequenceType],
             global_obs_shape: Union[int, SequenceType],
             action_shape: Union[int, SequenceType],
+            agent_num: int,
```
Comment (PaParaZz1) [inline ding/model/template/maqac.py:27]: remove this

File: dizoo/gfootball/config/gfootball_counter_mappo_config.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+evaluator_env_num = 32
+
+main_config = dict(
+    exp_name='gfootball_counter_mappo_seed0',
+    env=dict(
+        env_name='academy_counterattack_hard',
+        agent_num=agent_num,
+        obs_dim=obs_dim,
+        n_evaluator_episode=32,
+        stop_value=0.99,
```
Comment (PaParaZz1) [inline dizoo/gfootball/config/gfootball_counter_mappo_config.py:15]: win rate 1

File: dizoo/gfootball/config/gfootball_counter_mappo_config.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        collector_env_num=collector_env_num,
+        evaluator_env_num=evaluator_env_num,
+        manager=dict(
+            shared_memory=False,
+            reset_timeout=6000,
+        ),
+    ),
+    policy=dict(
+        cuda=True,
+        # share_weight=True,
```
Comment (PaParaZz1) [inline dizoo/gfootball/config/gfootball_counter_mappo_config.py:25]: remove this comment

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
+"""The code below is adapted from https://github.com/lich14/CDS/tree/main/CDS_GRF/envs/grf"""
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:1]: add unittest for this file

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+import gfootball.env as football_env
+from gfootball.env import observation_preprocessing
+import gym
+import numpy as np
+from ding.utils import ENV_REGISTRY
+from typing import Any, List, Union, Optional
+import copy
+import torch
+from ding.envs import BaseEnv, BaseEnvTimestep, BaseEnvInfo
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:10]: delete `BaseEnvInfo`, which is removed in main branch

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+    def get_global_special_state(self):
+        return [np.concatenate([self.get_global_state(), self.get_obs_agent(i)]) for i in range(self.n_agents)]
+
+    def check_if_done(self):
+        cur_obs = self._env.unwrapped.observation()[0]
+        ball_loc = cur_obs['ball']
+        ours_loc = cur_obs['left_team'][-self.n_agents:]
+
+        if ball_loc[0] < 0 or any(ours_loc[:, 0] < 0):
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:175]: add comment for this

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        """Returns the total number of actions an agent could ever take."""
+        return self.action_space[0].n
+
+    def render(self):
+        pass
+
+    def close(self):
+        self._env.close()
+
+    def save_replay(self):
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:277]: implement `save_replay` function

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    @property
+    def action_space(self) -> gym.spaces.Space:
+        return self._action_space
+
+    @property
+    def reward_space(self) -> gym.spaces.Space:
+        return self._reward_space
+
+    @staticmethod
+    def create_collector_env_cfg(cfg: dict) -> List[dict]:
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:304]: why add this method, `is_train` is not used in this file

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+    def get_avail_actions(self):
+        """Returns the available actions of all agents in a list."""
+        return [[1 for _ in range(self.n_actions)] for agent_id in range(self.n_agents)]
+
+    def get_avail_agent_actions(self, agent_id):
+        """Returns the available actions for agent_id."""
+        return self.get_avail_actions()[agent_id]
+
+    def get_total_actions(self):
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:267]: remove some methods like this, which is not used in this file now

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+            self._env.seed(self._seed)
+        self._final_eval_reward = 0
+
+        return obs
+
+    def step(self, actions):
+        """Returns reward, terminated, info."""
+        self.time_step += 1
+
+        if isinstance(actions, np.ndarray):
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:208]: remove torch requirement, all the elements in env should be `np.array` and python native scalar

File: dizoo/gfootball/envs/gfootball_academy_env.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+            dump_frequency=self.dump_freq,
+            number_of_left_players_agent_controls=self.n_agents,
+            number_of_right_players_agent_controls=self.number_of_right_players_agent_controls,
+            channel_dimensions=(observation_preprocessing.SMM_WIDTH, observation_preprocessing.SMM_HEIGHT)
+        )
+
+        obs_space_low = self._env.observation_space.low[0][:self.obs_dim]
+        obs_space_high = self._env.observation_space.high[0][:self.obs_dim]
+
+        self._action_space = [gym.spaces.Discrete(self._env.action_space.nvec[1]) for _ in range(self.n_agents)]
```
Comment (PaParaZz1) [inline dizoo/gfootball/envs/gfootball_academy_env.py:77]: update space like pettingzoo env

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/maqac.py (hunk 1 above)
Comment (puyuan1996) [reply ding/model/template/maqac.py:27]: 这个是为了和mavac一致，之后如果not share parameters时会用到的

File: dizoo/gfootball/envs/gfootball_academy_env.py (hunk 4 above)
Comment (puyuan1996) [reply dizoo/gfootball/envs/gfootball_academy_env.py:1]: 好的

File: dizoo/gfootball/envs/gfootball_academy_env.py (hunk 6 above)
Comment (puyuan1996) [reply dizoo/gfootball/envs/gfootball_academy_env.py:175]: 好的

File: dizoo/gfootball/config/gfootball_counter_mappo_config.py (hunk 3 above)
Comment (PaParaZz1) [reply dizoo/gfootball/config/gfootball_counter_mappo_config.py:25]: en

File: ding/model/template/maqac.py (hunk 1 above)
Comment (PaParaZz1) [reply ding/model/template/maqac.py:27]: yes

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**